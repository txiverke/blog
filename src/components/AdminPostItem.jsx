// @flow

import React from 'react'

import SingleInput from './form/SingleInput'
import Textarea from './form/Textarea'
import ImageUploader from './form/ImageUploader'
import { showFormErrors, showInputError } from '../utils/errorHandler'
import config from '../config'

type Props = {
  label: string,
  handlePost: Function,
  data?: Object
}

const AdminPostItem = ({ label, handlePost, data }: Props) => {
  
  const post = {}
  const hidden = data ? '' : 'hidden'

  function showContent (event: InputEvent) {
    const button = event.target.classList
    // $FlowFixMe
    const form = document.querySelector('.app-grid-body').classList

    if (button.contains('btn-hide')) {
      button.remove('btn-hide')
      button.add('btn-show')
      form.add('hidden')
    } else {
      button.remove('btn-show')
      button.add('btn-hide')
      form.remove('hidden')
    }
  }

  function handleChange (event: InputEvent) {
    event.target.classList.add('active')
    showInputError(event.target)
  }

  function handleData (event: InputEvent) {
    const title = event.target.elements.title.value.trim()
    const content = event.target.elements.content.value.trim()
    const link = event.target.elements.link.value.trim()
    const tags = event.target.elements.tags.value.trim()

    return Object.assign(post, { title, content, link, tags })
  }

  function handleImageChange (file: File) {
    if (document.querySelector('.app-preview') && file) {
      // $FlowFixMe
      document.querySelector('.app-preview').classList.add('hidden')
    }

    return Object.assign(post, { file })
  } 
  
  return (
    <article className="app-grid"> 
      <header className="app-grid-header">
        <h2 className="tit-section">{label}</h2>
        {!data && 
          <button 
            onClick={showContent}
            className="btn btn-icon btn-clean icon-close btn-show tr2">
          </button>
        }
      </header>
      <form
        noValidate
        encType="multipart/form-data"
        className={`app-grid-body ${hidden}`} 
        onSubmit={
          (event) => {
            event.preventDefault()
            if (showFormErrors()) {
              let id = data ? data._id : ''
              handlePost(handleData(event), id)
            }
        }}
      >
        <SingleInput
          wrapper="app-grid-whole"
          name="title"
          inputType="text"
          title="Title"
          placeholder="Title"
          content={data && data.title}
          pattern=".{6,}"
          controlFunc={handleChange}
        />
        <SingleInput
          wrapper="app-grid-item1"
          name="link"
          inputType="text"
          title="Link"
          placeholder="Link"
          content={data && data.link}
          pattern=".{6,}"
          controlFunc={handleChange}
        />
        <SingleInput
          wrapper="app-grid-item2"
          name="tags"
          inputType="text"
          title="Tags"
          placeholder="Tags"
          content={data && data.tags}
          pattern=".{2,}"
          controlFunc={handleChange}
        />
        <ImageUploader 
          handleImage={file => handleImageChange(file)}
        />
        {data && !post.file &&
          <img 
            className="app-preview"
            src={`${config.api.public}/posts/${data.background}`} 
            height="200" 
            alt="Preview..." 
          />
        }
        <Textarea
          wrapper="app-grid-whole"
          name="content"
          title="Content"
          pattern=".{6,}"
          content={data && data.content}
          controlFunc={handleChange}
        />
        <button 
          type="submit"
          className="app-grid-btn btn"
        >
        {label}
        </button>
      </form>
    </article>
  )
}

export default AdminPostItem