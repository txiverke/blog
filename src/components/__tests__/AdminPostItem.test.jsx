// @flow

import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import AdminPostItem from '../AdminPostItem'

let properties = {
  label: 'test label',
  handlePost: jest.fn(),
  data: {}
}

function setup () {
  const props = properties

  const enzymeWrapper = mount(
    <MemoryRouter initialEntries={['/test']}>
      <AdminPostItem {...props} />
    </MemoryRouter>
  ) 

  return {
    props,
    enzymeWrapper
  }
}

describe('<AdminPostItem />', () => {
  it('should render itself', () => {
    const component = shallow(<AdminPostItem {...properties} />)
    expect(component).toMatchSnapshot()
  })

  it('should not render the toogle button if there is data', () => {
    properties.data = { test: true }
    const component = shallow(<AdminPostItem {...properties} />)
    expect(component.find('header > .btn').length).toEqual(0)
  })

  it('should sent the data onSubmit the form', () => {
    const { enzymeWrapper, props } = setup()
    expect(props.handlePost.mock.calls.length).toEqual(0)
    enzymeWrapper.find('form').simulate('submit')
    expect(props.handlePost.mock.calls.length).toEqual(1)
  })

})