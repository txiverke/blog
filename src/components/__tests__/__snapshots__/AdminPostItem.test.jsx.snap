// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminPostItem /> should render itself 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminPostItem
    data={Object {}}
    handlePost={[Function]}
    label="test label"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="app-grid-header"
>
          <h2
                    className="tit-section"
          >
                    test label
          </h2>
          <button
                    className="btn btn-icon btn-clean icon-close btn-show tr2"
                    onClick={[Function]}
          />
</header>,
        <form
          className="app-grid-body "
          encType="multipart/form-data"
          noValidate={true}
          onSubmit={[Function]}
>
          <SingleInput
                    content=""
                    controlFunc={[Function]}
                    inputType="text"
                    name="title"
                    pattern=".{6,}"
                    placeholder="Title"
                    required={true}
                    size=""
                    title="Title"
                    wrapper="app-grid-whole"
          />
          <SingleInput
                    content=""
                    controlFunc={[Function]}
                    inputType="text"
                    name="link"
                    pattern=".{6,}"
                    placeholder="Link"
                    required={true}
                    size=""
                    title="Link"
                    wrapper="app-grid-item1"
          />
          <SingleInput
                    content=""
                    controlFunc={[Function]}
                    inputType="text"
                    name="tags"
                    pattern=".{2,}"
                    placeholder="Tags"
                    required={true}
                    size=""
                    title="Tags"
                    wrapper="app-grid-item2"
          />
          <ImageUploader
                    handleImage={[Function]}
          />
          <img
                    alt="Preview..."
                    className="app-preview"
                    height="200"
                    src="http://localhost:8888/blog/img/posts/undefined"
          />
          <TextArea
                    content=""
                    controlFunc={[Function]}
                    name="content"
                    pattern=".{6,}"
                    placeholder=""
                    required={true}
                    resize={false}
                    rows={5}
                    title="Content"
                    wrapper="app-grid-whole"
          />
          <button
                    className="app-grid-btn btn"
                    type="submit"
          >
                    test label
          </button>
</form>,
      ],
      "className": "app-grid",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2
              className="tit-section"
>
              test label
</h2>,
            <button
              className="btn btn-icon btn-clean icon-close btn-show tr2"
              onClick={[Function]}
/>,
          ],
          "className": "app-grid-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "test label",
              "className": "tit-section",
            },
            "ref": null,
            "rendered": "test label",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "btn btn-icon btn-clean icon-close btn-show tr2",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <SingleInput
              content=""
              controlFunc={[Function]}
              inputType="text"
              name="title"
              pattern=".{6,}"
              placeholder="Title"
              required={true}
              size=""
              title="Title"
              wrapper="app-grid-whole"
/>,
            <SingleInput
              content=""
              controlFunc={[Function]}
              inputType="text"
              name="link"
              pattern=".{6,}"
              placeholder="Link"
              required={true}
              size=""
              title="Link"
              wrapper="app-grid-item1"
/>,
            <SingleInput
              content=""
              controlFunc={[Function]}
              inputType="text"
              name="tags"
              pattern=".{2,}"
              placeholder="Tags"
              required={true}
              size=""
              title="Tags"
              wrapper="app-grid-item2"
/>,
            <ImageUploader
              handleImage={[Function]}
/>,
            <img
              alt="Preview..."
              className="app-preview"
              height="200"
              src="http://localhost:8888/blog/img/posts/undefined"
/>,
            <TextArea
              content=""
              controlFunc={[Function]}
              name="content"
              pattern=".{6,}"
              placeholder=""
              required={true}
              resize={false}
              rows={5}
              title="Content"
              wrapper="app-grid-whole"
/>,
            <button
              className="app-grid-btn btn"
              type="submit"
>
              test label
</button>,
          ],
          "className": "app-grid-body ",
          "encType": "multipart/form-data",
          "noValidate": true,
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "content": "",
              "controlFunc": [Function],
              "inputType": "text",
              "name": "title",
              "pattern": ".{6,}",
              "placeholder": "Title",
              "required": true,
              "size": "",
              "title": "Title",
              "wrapper": "app-grid-whole",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "content": "",
              "controlFunc": [Function],
              "inputType": "text",
              "name": "link",
              "pattern": ".{6,}",
              "placeholder": "Link",
              "required": true,
              "size": "",
              "title": "Link",
              "wrapper": "app-grid-item1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "content": "",
              "controlFunc": [Function],
              "inputType": "text",
              "name": "tags",
              "pattern": ".{2,}",
              "placeholder": "Tags",
              "required": true,
              "size": "",
              "title": "Tags",
              "wrapper": "app-grid-item2",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "handleImage": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "Preview...",
              "className": "app-preview",
              "height": "200",
              "src": "http://localhost:8888/blog/img/posts/undefined",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "content": "",
              "controlFunc": [Function],
              "name": "content",
              "pattern": ".{6,}",
              "placeholder": "",
              "required": true,
              "resize": false,
              "rows": 5,
              "title": "Content",
              "wrapper": "app-grid-whole",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "test label",
              "className": "app-grid-btn btn",
              "type": "submit",
            },
            "ref": null,
            "rendered": "test label",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="app-grid-header"
>
            <h2
                        className="tit-section"
            >
                        test label
            </h2>
            <button
                        className="btn btn-icon btn-clean icon-close btn-show tr2"
                        onClick={[Function]}
            />
</header>,
          <form
            className="app-grid-body "
            encType="multipart/form-data"
            noValidate={true}
            onSubmit={[Function]}
>
            <SingleInput
                        content=""
                        controlFunc={[Function]}
                        inputType="text"
                        name="title"
                        pattern=".{6,}"
                        placeholder="Title"
                        required={true}
                        size=""
                        title="Title"
                        wrapper="app-grid-whole"
            />
            <SingleInput
                        content=""
                        controlFunc={[Function]}
                        inputType="text"
                        name="link"
                        pattern=".{6,}"
                        placeholder="Link"
                        required={true}
                        size=""
                        title="Link"
                        wrapper="app-grid-item1"
            />
            <SingleInput
                        content=""
                        controlFunc={[Function]}
                        inputType="text"
                        name="tags"
                        pattern=".{2,}"
                        placeholder="Tags"
                        required={true}
                        size=""
                        title="Tags"
                        wrapper="app-grid-item2"
            />
            <ImageUploader
                        handleImage={[Function]}
            />
            <img
                        alt="Preview..."
                        className="app-preview"
                        height="200"
                        src="http://localhost:8888/blog/img/posts/undefined"
            />
            <TextArea
                        content=""
                        controlFunc={[Function]}
                        name="content"
                        pattern=".{6,}"
                        placeholder=""
                        required={true}
                        resize={false}
                        rows={5}
                        title="Content"
                        wrapper="app-grid-whole"
            />
            <button
                        className="app-grid-btn btn"
                        type="submit"
            >
                        test label
            </button>
</form>,
        ],
        "className": "app-grid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2
                className="tit-section"
>
                test label
</h2>,
              <button
                className="btn btn-icon btn-clean icon-close btn-show tr2"
                onClick={[Function]}
/>,
            ],
            "className": "app-grid-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "test label",
                "className": "tit-section",
              },
              "ref": null,
              "rendered": "test label",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "btn btn-icon btn-clean icon-close btn-show tr2",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": "button",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <SingleInput
                content=""
                controlFunc={[Function]}
                inputType="text"
                name="title"
                pattern=".{6,}"
                placeholder="Title"
                required={true}
                size=""
                title="Title"
                wrapper="app-grid-whole"
/>,
              <SingleInput
                content=""
                controlFunc={[Function]}
                inputType="text"
                name="link"
                pattern=".{6,}"
                placeholder="Link"
                required={true}
                size=""
                title="Link"
                wrapper="app-grid-item1"
/>,
              <SingleInput
                content=""
                controlFunc={[Function]}
                inputType="text"
                name="tags"
                pattern=".{2,}"
                placeholder="Tags"
                required={true}
                size=""
                title="Tags"
                wrapper="app-grid-item2"
/>,
              <ImageUploader
                handleImage={[Function]}
/>,
              <img
                alt="Preview..."
                className="app-preview"
                height="200"
                src="http://localhost:8888/blog/img/posts/undefined"
/>,
              <TextArea
                content=""
                controlFunc={[Function]}
                name="content"
                pattern=".{6,}"
                placeholder=""
                required={true}
                resize={false}
                rows={5}
                title="Content"
                wrapper="app-grid-whole"
/>,
              <button
                className="app-grid-btn btn"
                type="submit"
>
                test label
</button>,
            ],
            "className": "app-grid-body ",
            "encType": "multipart/form-data",
            "noValidate": true,
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "content": "",
                "controlFunc": [Function],
                "inputType": "text",
                "name": "title",
                "pattern": ".{6,}",
                "placeholder": "Title",
                "required": true,
                "size": "",
                "title": "Title",
                "wrapper": "app-grid-whole",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "content": "",
                "controlFunc": [Function],
                "inputType": "text",
                "name": "link",
                "pattern": ".{6,}",
                "placeholder": "Link",
                "required": true,
                "size": "",
                "title": "Link",
                "wrapper": "app-grid-item1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "content": "",
                "controlFunc": [Function],
                "inputType": "text",
                "name": "tags",
                "pattern": ".{2,}",
                "placeholder": "Tags",
                "required": true,
                "size": "",
                "title": "Tags",
                "wrapper": "app-grid-item2",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "handleImage": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "Preview...",
                "className": "app-preview",
                "height": "200",
                "src": "http://localhost:8888/blog/img/posts/undefined",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "content": "",
                "controlFunc": [Function],
                "name": "content",
                "pattern": ".{6,}",
                "placeholder": "",
                "required": true,
                "resize": false,
                "rows": 5,
                "title": "Content",
                "wrapper": "app-grid-whole",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "test label",
                "className": "app-grid-btn btn",
                "type": "submit",
              },
              "ref": null,
              "rendered": "test label",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
